File: frontend/packages/types/src/subscription.ts
export type MembershipPlanCode = "free" | "plus" | "premium";
export type MembershipBillingInterval = "monthly" | "annual";
export type MembershipStatus =
  | "active"
  | "canceled"
  | "past_due"
  | "trialing"
  | "incomplete";

export type SubscriptionPlanFeature = {
  key: string;
  label: string;
  included: boolean;
};

export type SubscriptionPlanPrice = {
  monthly_cents: number;
  annual_cents: number;
  annual_monthly_equivalent_cents: number;
  currency: string;
};

export type SubscriptionPlan = {
  code: MembershipPlanCode;
  name: string;
  tagline: string;
  display_order: number;
  is_highlighted: boolean;
  price: SubscriptionPlanPrice;
  features: SubscriptionPlanFeature[];
};

export type SubscriptionPlansResponse = {
  plans: SubscriptionPlan[];
  annual_discount_percent: number;
  checkout_enabled: boolean;
};

export type SubscriptionMe = {
  plan_code: MembershipPlanCode;
  billing_interval: MembershipBillingInterval | null;
  status: MembershipStatus;
  is_paid: boolean;
  current_period_end: string | null;
  can_upgrade: boolean;
};

export type SubscriptionSupporterAvatar = {
  display_name: string;
  avatar_url: string | null;
};

export type SubscriptionCommunityStats = {
  supporter_count: number;
  avatars: SubscriptionSupporterAvatar[];
};

export type SubscriptionCheckoutRequest = {
  plan_code: "plus" | "premium";
  billing_interval: MembershipBillingInterval;
};

export type SubscriptionCheckoutResponse = {
  status: "redirect" | "unavailable";
  checkout_url: string | null;
  message: string | null;
};

File: backend/app/schemas/subscription.py
"""Membership subscription API schemas."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from app.core.subscription_constants import (
    MembershipBillingInterval,
    MembershipPlanCode,
    MembershipStatus,
)


class SubscriptionPlanFeatureResponse(BaseModel):
    key: str
    label: str
    included: bool


class SubscriptionPlanPriceResponse(BaseModel):
    monthly_cents: int
    annual_cents: int
    annual_monthly_equivalent_cents: int
    currency: str = "EUR"


class SubscriptionPlanResponse(BaseModel):
    code: MembershipPlanCode
    name: str
    tagline: str
    display_order: int
    is_highlighted: bool
    price: SubscriptionPlanPriceResponse
    features: list[SubscriptionPlanFeatureResponse]


class SubscriptionPlansResponse(BaseModel):
    plans: list[SubscriptionPlanResponse]
    annual_discount_percent: int
    checkout_enabled: bool


class SubscriptionMeResponse(BaseModel):
    plan_code: MembershipPlanCode
    billing_interval: MembershipBillingInterval | None = None
    status: MembershipStatus
    is_paid: bool
    current_period_end: datetime | None = None
    can_upgrade: bool


class SubscriptionSupporterAvatarResponse(BaseModel):
    display_name: str
    avatar_url: str | None = None


class SubscriptionCommunityStatsResponse(BaseModel):
    supporter_count: int
    avatars: list[SubscriptionSupporterAvatarResponse]


class SubscriptionCheckoutRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    plan_code: MembershipPlanCode = Field(..., description="plus or premium")
    billing_interval: MembershipBillingInterval = MembershipBillingInterval.MONTHLY


class SubscriptionCheckoutStatus(StrEnum):
    REDIRECT = "redirect"
    UNAVAILABLE = "unavailable"


class SubscriptionCheckoutResponse(BaseModel):
    status: SubscriptionCheckoutStatus
    checkout_url: str | None = None
    message: str | None = None
